### DISCUSSION TITLE
How to troubleshoot recipe application (Maven-based execution)

### DISCUSSION THREAD

**Discussion by crankydillo** [openrewrite/rewrite#4256] [Q&A] [answered]
Is there any way to enable logging so that what's going on with recipe application can be understood?  Are there any best practices in this area?  Apologies if I missed it in the doc!

**Comment by timtebeek** [comment]
hi! What aspect are you looking to troubleshoot?

If you're finding it's taking a long time before it starts to run a recipe, then we might just be building up the model of your code, as explained here: https://docs.openrewrite.org/reference/faq#my-recipe-appears-to-hang-when-running.-whats-happening-is-there-a-progress-report

If you're looking to attach a debugger, that's possible through the usual ways for Maven using `mvnDebug`. That could help drop you into a recipe visitor if you set a corresponding breakpoint there. It's not explicitly documented as it's not unique to OpenRewrite, but perhaps it would help to call this out in the docs (cc @mike-solomon ). We did cover this in [last week's live stream](https://www.youtube.com/watch?v=lNLo6i7SVGI&t=1991s), in case that helps you there. For Gradle, the steps are documented in https://github.com/openrewrite/rewrite/issues/4054#issuecomment-2119019047

Feel free to respond with more details as to what you're after, or mark this as the answer if it helps you out there.

**Reply by crankydillo** [reply]
I have been using `mvnDebug`, but I was really hoping for logging.  Something that would show which recipe was getting processed and why it may not apply to a particular file.  Stuff like that.

Update: I'm referring to recipes in a composite recipe and knowing when and why one is not having the intended effect.

**Reply by timtebeek** [reply, marked as answer]
That's helpful context, thanks! We've opted not to produce logging for such insights, but to produce data tables instead: https://docs.openrewrite.org/running-recipes/data-tables

Logging can get quite noisy, especially when you multiple ~2000 recipes for a Spring Boot migration times the number of files in a project. It's also not ideal for a structured analysis to gain the insights that you're after.

So instead we [produce three data tables by default](https://github.com/openrewrite/rewrite-recipe-markdown-generator/blob/208a9d193b6fbcc06e5bcec3b6a11b425b0e4bb5/src/main/kotlin/org/openrewrite/RecipeMarkdownGenerator.kt#L876-L880) for every recipe run, if enabled, such that you can get insight into which recipes had results, or errors, and their runtime statistics. Look for these tables in your target folder to learn which recipes had results.
```
"org.openrewrite.table.SourcesFileResults",
"org.openrewrite.table.SourcesFileErrors",
"org.openrewrite.table.RecipeRunStats"
```

**Reply by crankydillo** [reply]
I enabled that 5.33.0 and only see RecipeRunStats.

```
mvn -e -Drewrite.exportDatatables=true -Drewrite.runPerSubmodule=false -Dmaven.test.skip=true -Denforcer.skip=true compile -Drewrite.recipeArtifactCoordinates= ...
```

I'm hopeful 'SourcesFileErrors' would be useful.  Debugging led me to a problem downloading Maven, and I eventually found some code that made me think using `-X` would show info.  It showed:

```
[DEBUG]
    19 org.openrewrite.maven.MavenDownloadingExceptions
    18     at org.openrewrite.maven.MavenDownloadingExceptions.append (MavenDownloadingExceptions.java:44)
    17     at org.openrewrite.maven.tree.MavenResolutionResult.resolveDependencies (MavenResolutionResult.java:180)
    16     at org.openrewrite.maven.MavenParser.parseInputs (MavenParser.java:116)
    15     at org.openrewrite.Parser.parse (Parser.java:59)
    14     at org.openrewrite.maven.MavenMojoProjectParser.parseMaven (MavenMojoProjectParser.java:557)
    13     at org.openrewrite.maven.MavenMojoProjectParser.listSourceFiles (MavenMojoProjectParser.java:149)
...
7     Suppressed: org.openrewrite.maven.MavenDownloadingException: com.graphql-java-kickstart:graphql-java-tools failed. No version provided
     6         at org.openrewrite.maven.tree.ResolvedPom.resolveDependencies (ResolvedPom.java:834)
     5         at org.openrewrite.maven.tree.ResolvedPom.resolveDependencies (ResolvedPom.java:806)
     4         at org.openrewrite.maven.tree.MavenResolutionResult.resolveDependencies (MavenResolutionResult.java:175)
     3         at org.openrewrite.maven.MavenParser.parseInputs (MavenParser.java:116)
```

Don't really need to solve this (I suspect it may have something to do with a parent pom's dependency management section having an import-scope dependency).  I'm really just interested in how to quickly triage this stuff.

I will try to figure out why I'm not seeing `SourcesFileErrors`.  To triage, is that the best I have?  Or maybe use `mvn -X`?  If those fail,  attach a debugger?

**Reply by timtebeek** [reply]
As to those downloading exceptions we learned that we might be trying to resolve some dependencies unnecessarily as compared to Maven, for instance when they are a transitive dependency in the provided scope. That could be a source of such downloading exceptions that we hope to resolve shortly.

The tables with results and errors would only be present if there are errors or results of course. They won't inform you of files not matched or modified, as again: that's likely to produce a lot of results where a debugger might prove more valuable to troubleshoot.

It might all come down to what kind of errors you're phasing, and how best to debug those; common debug options that might help are these recipes that each produce a data table hinting at possible issues:
- https://docs.openrewrite.org/recipes/core/findparsefailures
- https://docs.openrewrite.org/recipes/java/search/findmissingtypes

**Reply by crankydillo** [reply]
Thank you @timtebeek .  We'll work with these and get back to you if we have more questions/thoughts.

**Reply by timtebeek** [reply]
All good! I've produced a short write up in the docs using some of the above information, such that it should be easier to find going forward: https://docs.openrewrite.org/reference/faq#my-recipe-runs-but-is-not-making-any-changes.-whats-happening
Hope that helps!

**Reply by crankydillo** [reply]
Thanks for that FAQ @timtebeek .

I think just one more question in this area (at least for now:). Do you think these would be useful to generally run with in order to easily spot a problem?  They aren't in your FAQ entry, but look like triage-y things:)

```
 org.openrewrite.java.dependencies.DependencyResolutionDiagnostic
 org.openrewrite.FindDeserializationErrors
```

**Reply by timtebeek** [reply]
Yes great addition that first one! Added just now: https://github.com/openrewrite/rewrite-docs/commit/93527d3670b516c350faca901de7a0011e9e1cd2

The `FindDeserializationErrors` should only apply to Moderne users; I don't think that'll be common enough to list in the OpenRewrite docs.

**Reply by crankydillo** [reply]
Thanks!

**Comment by crankydillo** [comment]
Don't need to discuss it here, but I believe this is [the issue](https://github.com/openrewrite/rewrite/issues/4269) we are facing.
